### PR TITLE
quantis, multi_engine and xtc compression (gromacs)

### PR DESCRIPTION
**Pull request by lukasbaldauf**
Features added:
## ensemble_engines
Added an option `ensemble_engines` in the [simulation] section, where we can specify one or multiple engines for each ensemble. The global dictionary ENGINES now contains a nested list for each unique engine. This nested list has two items; an occupation list (what workers currently are using what engine), and a list of the specific engine instances. We now need the additional information on what engines are in-use because we can have fewer engines than workers. For example if we have a single different engine __engineX__ in one of the ensembles, and all other engines are identical, we only initiate one instance of __engineX__. 

## .xtc compression
When studying large systems we might want to store compressed coordinates in the .xtc format by settings `nstxout-compressed` in the gromacs.mdp. Using `keep_traj_fnames = [".xtc"]` in the [output] section of the .toml now moves any additional trajectory files that are written (with the same basename, e.g. 000_1234_trajF.xtc) into the accepted/ folder. They will not be deleted even if `delete_old = true`. This can be used for any engine that writes additional files. For example, we can use `keep_traj_fnames =  ["-frc-1.xyz"]` for cp2k forces. We can also specify `compressed-x-grps` in the .mdp to select a subset of atoms, or increase/ decrease the frequency of the compression with the `nstxout-compressed` keyword in the .mdp file, allowing for higher or lower than `subcycles` output. Note: if `nstxout-compressed` is higher than `subcycles`, the first, final, and some intermediate phase points might be missing from the trajectories.

## quantis
The quantis method can now be used with `quantis = true` in the [tis_set] section of the .toml. It requires an additional [engine0] section. We can pass the keyword `accept_all = true` in the [tis_set] section to accept all zero swaps.

## other stuff
Added a function that checks the config dictionary created from the .toml.

### CODE REVIEW

File: infretis/classes/engines/turtlemdengine.py
Comment (lukasbaldauf) [inline infretis/classes/engines/turtlemdengine.py:1]: test

File: infretis/classes/formatter.py
Comment (lukasbaldauf) [inline infretis/classes/formatter.py:1]: Here we need to return a dictionary of trajectory names so we can access it when moving the trjectori_i.xtc to accepted/ later.

File: infretis/core/tis.py
Comment (lukasbaldauf) [inline infretis/core/tis.py:1]: We now select the specific engine in the `select_shoot()` function, because the engine can depend on the MC move.

**Comment by lukasbaldauf** [comment]
Doesn't seem to run with external orderparameters (and other external modules), e.g. `examples/turtlemd/double_well`

File: infretis/core/tis.py [hunk 1]
```diff
@@ -279,9 +276,25 @@ def select_shoot(
...
+            engine = ENGINES[-1]
+        elif pens["ens"]["tis_set"]["multi_engine"]:
+            engine = ENGINES[ens_num]
+        else:
+            engine = ENGINES[picked[ens_num]["pin"]]
+        engine.set_mdrun(pens)
+        if "rgen-eng" in pens:
+            engine.rgen = pens["rgen-eng"]
+        engine.clean_up()
+        pens["ens"]["engine"] = engine
```
Comment (lukasbaldauf) [inline infretis/core/tis.py:292]: It was this line that created the bug. I am not sure why, but maybe adding the engine to the `pens` dict makes the engine get pickled.

File: infretis/classes/engines/enginebase.py
Comment (lukasbaldauf) [inline infretis/classes/engines/enginebase.py:1]: Add beta property so an error is thrown if the engine doesn't define it.

File: infretis/classes/formatter.py
Comment (lukasbaldauf) [inline infretis/classes/formatter.py:1]: Change formatter class to move certain files from workerX/ to load/i/accepted/

**Comment by dz24** [comment]
What about changing "multi_engine" in toml.simulation.tis_set to "engines" in toml.simulation? Where we set a default to "engines = ["engine1"]", and then we have one additional variable "ens_engs = ["engine1", "engine1", "engine1"..] for each ensemble. Then it becomes straightforward to change into quantis (or other multi-engine schemes) as it will be similar to changing shooting moves in ensembles. Plus we do not have to duplicate the same engine multiple times.

Also, I think it would be nice to add another heavy test in infretis/test/simulations for quantis, where an actual simulation is ran. Through github actions we can ensure that we get same quantis output over time etc. 

@lukasbaldauf

**Comment by lukasbaldauf** [comment]
Yes that would make it easier! In the case of multiple engines per ensemble, would you do something like “ens_eng = [[‘eng0’, ‘eng1’], ‘eng2’, ‘eng2’,…], a nested list? 

Also, for the keyword “engines”, would it be enough to find all unique entries of the list “ens_eng” to infer the “engines” value? Or does “engines” specify some other information I’m not seeing? 

I’ll add a heavy test as well.

**Comment by dz24** [comment]
> Yes that would make it easier! In the case of multiple engines per ensemble, would you do something like “ens_eng = [[‘eng0’, ‘eng1’], ‘eng2’, ‘eng2’,…], a nested list?

Sure, this will probably work.
 
> Also, for the keyword “engines”, would it be enough to find all unique entries of the list “ens_eng” to infer the “engines” value? Or does “engines” specify some other information I’m not seeing?

If we only have "ens_eng", then it may be burdensome for the user to repeat same engine in all ensembles if they don't run quantis. I originally wanted the "engines" variable so that if the user does not also define "ens_eng" in toml, it get automatically filled with the same engine in all ensembles.

But, maybe we can remove both "engines" and "ens_eng" variables in .toml when we run no quantis (as we do right now). In the case of no ens_eng in .toml, it get filled automatically with same engines. If we do run quantis, we must write "ens_eng" variables in .toml, which must have a size equal number of ensembles.

**Comment by lukasbaldauf** [comment]
It’s a bit tricky with the ens_eng tho, because in some case (ASE engine and also AMS I believe) we have to specify the hardware the engine runs on. The engine will then be initiated once answering kept alive for the whole simulation. If we then have 1 engine in each ensemble instead of 1 engine per worker, we have to allocate more resources, half of which will not be used because the ensembles are free. That’s what I initially tried to avoid.

**Comment by dz24** [comment]
> If we then have 1 engine in each ensemble

The actual number of engines to be initiated can be different, just for the .toml file I think we should write the engine(s) to be used in each ensemble. But yes, I did not consider that allocating hardware may be more difficult with different types of engines. I think you can decide how things should look.

**Comment by lukasbaldauf** [comment]
I think I’ll go for something with the ens_eng list. It’s pretty annoying to define a huge nr. of [engine] sections in the .toml. In general, we probably don’t have as many unique engines as ensembles.

**Comment by lukasbaldauf** [comment]
I added the `ensemble_engines` option now. It made everything a bit neater wrt. quantis.

I haven't had time to add another simulation quantis test, but i added a heavy test for the zero swaps earlier, which is the only difference between the two. Ill try to add another quantis test when i have more time after next week. I need to make another tutorial for the lammps engine before that.

**Comment by dz24** [comment]
> [call def_globals in setup to avoid creating engines multiple times](https://github.com/infretis/infretis/pull/109/commits/e462c81aa5a5fca50e45fc46824d86c32330f854)

Does this work for you? Setting def_globals in setup results my simulations to crash in the beginning, as the ENGINES object is empty / {}. Maybe it's better if we just make engines local again?

**Comment by lukasbaldauf** [comment]
Hmm, it works for me, and also the tests that run infretis pass. Maybe its a mac issue then? Something with global variables across modules? I haven't really looked into the global variable stuff and if it is needed, but I noticed that the ENGINES dict was initiated n_worker times at some point. I also noticed that stuff still gets pickled, so I didn't want to make engines local again because I was afraid it would reintroduce the errors we initially tried to avoid.

What system is it?

**Comment by lukasbaldauf** [comment]
Does something like this work

**Comment by dz24** [comment]
No. But I know what the problem is, it is as esclapes said, https://github.com/infretis/infretis/pull/103#issuecomment-2319098646, in windows/macos, childprocesses are spawned while for linux they are forked. 

https://stackoverflow.com/a/65221779/21385538

This fixed it for me, in asyncrunner.py,

        self._executor: concurrent.futures.Executor = (
            concurrent.futures.ProcessPoolExecutor(
                max_workers=n_workers,
                initializer=worker_initializer,
                initargs=(self._counter,),
                mp_context=multiprocessing.get_context('fork'),
            )
        )

**Comment by dz24** [comment]
It works for me now.

**Review by dz24 — APPROVED** [review]
LGTM
